File: mini_utils/WTimer.cpp

#include "Log.h"
#include "Macro.h"

#include "WTimer.h"

using namespace MSA;

//第1个轮子槽0~255
#define TVR_MSAK (TVR_SIZE - 1)

//第2~5轮子槽0~63
#define TVN_MSAK (TVN_SIZE - 1)


WTimer::WTimer(uint16_t tickSize)
{
	m_Wheel.curTick = 0;
	for (auto i = 0; i < TVR_SIZE; i++)
	{
		auto entry = m_Wheel.tv1 + i;
		entry->head = entry->tail = nullptr;
	}

	for (auto i = 0; i < TVN_SIZE; i++)
	{
		auto entry = m_Wheel.tv2 + i;
		entry->head = entry->tail = nullptr;
		entry = m_Wheel.tv3 + i;
		entry->head = entry->tail = nullptr;
		entry = m_Wheel.tv4 + i;
		entry->head = entry->tail = nullptr;
		entry = m_Wheel.tv5 + i;
		entry->head = entry->tail = nullptr;

	}
	m_TickSize = tickSize < 1 ? 1 : tickSize;
}

WTimer::~WTimer()
{
	DelTvTimer(m_Wheel.tv1, TVR_SIZE);
	DelTvTimer(m_Wheel.tv2, TVN_SIZE);
	DelTvTimer(m_Wheel.tv3, TVN_SIZE);
	DelTvTimer(m_Wheel.tv4, TVN_SIZE);
	DelTvTimer(m_Wheel.tv5, TVN_SIZE);
	if (nullptr != m_CacheTimerArray)
	{
		SAFE_DELETE_ARRAY(m_CacheTimerArray);
	}
}

void WTimer::DelTvTimer(List* tv, uint16_t size)
{
	for (auto i = 0; i < size; i++)
	{
		List* entry = (tv + i);
		auto curr = entry->head;
		while (nullptr != curr)
		{
			Timer* next = curr->next;
			if (curr->id > m_CacheTimerNum)
			{
				SAFE_DELETE(curr);
			}
			curr = next;
		}
		entry->head = entry->tail = nullptr;
	}
}


bool WTimer::Init(uint32_t cacheTimerNum)
{
	//已初始化
	if (m_Wheel.curTick > 0) return false;
	auto nowTime = SysTime::GetSysMilSecond();
	m_Wheel.curTick = nowTime / m_TickSize;

	if (0 == cacheTimerNum)  return false;
	if (cacheTimerNum < 1)
	{
		m_CacheTimerNum = 1024;
	}
	else
	{
		m_CacheTimerNum = cacheTimerNum;
	}

	return InitCacheTimer(m_CacheTimerNum);
}

uint32_t WTimer::AddTimer(const std::function<void(uint64_t)> fun, uint64_t data, uint32_t delay, uint32_t interval, int16_t runNum)
{
	auto timer = (Timer*)m_CustomStack.Pop();
	if (nullptr == timer)
	{
		timer = new(nothrow) Timer;
		if (nullptr == timer)
		{
			LogError("not enough memory");
			return 0;
		}
		timer->id = GetNewTimerId();
	}
	timer->fun = fun;
	timer->data = data;
	timer->runNum = runNum == 0 ? 1 : runNum;
	timer->interval = interval < 0 ? 1 : interval;

	if (delay < 0) delay = 0;
	timer->expire = SysTime::GetSysMilSecond() + delay;
	m_TimerMap.insert(pair<uint32_t, Timer*>(timer->id, timer));

	AddTimer(timer); return timer->id;

}

void WTimer::AddTimer(Timer* timer)
{
	//过期时刻度
	auto ticks = timer->expire / m_TickSize;
	uint64_t idx = ticks - m_Wheel.curTick;

	List* entry;
	auto i = 0;
	if (idx < TVR_SIZE) {
		i = ticks & TVR_MSAK;
		entry = m_Wheel.tv1 + i;
	}
	else if (idx < 1 << (TVR_BITS + TVN_BITS)) {
		i = (ticks >> TVR_BITS) & TVN_MSAK;
		entry = m_Wheel.tv2 + i;
	}
	else if (idx < 1 << (TVR_BITS + 2 * TVN_BITS)) {
		i = (ticks >> (TVR_BITS + TVN_BITS)) & TVN_MSAK;
		entry = m_Wheel.tv3 + i;
	}
	else if (idx < 1 << (TVR_BITS + 3 * TVN_BITS)) {
		i = (ticks >> (TVR_BITS + 2 * TVN_BITS)) & TVN_MSAK;
		entry = m_Wheel.tv4 + i;
	}
	else if ((signed long)idx < 0) {
		entry = m_Wheel.tv1 + (m_Wheel.curTick & TVR_MSAK);
	}
	else {
		if (idx > 0xffffffffUL) {
			idx = 0xffffffffUL;
			ticks = idx + m_Wheel.curTick;
		}
		i = (ticks >> (TVR_BITS + 3 * TVN_BITS)) & TVN_MSAK;
		entry = m_Wheel.tv5 + i;
	}
	if (nullptr != entry->tail)
	{
		timer->prev = entry->tail;
		entry->tail->next = timer;
		entry->tail = timer;
		timer->entry = entry;
		timer->next = nullptr;
	}
	else
	{
		timer->entry = entry;
		entry->head = entry->tail = timer;
		timer->prev = timer->next = nullptr;
	}

}


void WTimer::Scheduled()
{
	Scheduled(SysTime::GetSysMilSecond());
}

uint8_t WTimer::Cascade(List* tv, uint8_t idx)
{
	auto entry = tv + idx;
	auto curr = entry->head;
	entry->head = nullptr;
	entry->tail = nullptr;
	while (nullptr != curr)
	{
		auto next = curr->next;
		AddTimer(curr); curr = next;
	}
	return idx;
}

#define INDEX(N) ((m_Wheel.curTick >> (TVR_BITS + (N) * TVN_BITS)) & TVN_MSAK)

//currTime:调度时间要大于StartTime单位(毫秒)
void WTimer::Scheduled(uint64_t currTime)
{
	auto currTick = currTime / m_TickSize;
	while (m_Wheel.curTick < currTick)
	{
		auto idx = m_Wheel.curTick & TVR_MSAK;

		if (!idx &&
			(!Cascade(m_Wheel.tv2, INDEX(0))) &&
			(!Cascade(m_Wheel.tv3, INDEX(1))) &&
			!Cascade(m_Wheel.tv4, INDEX(2)))
		{
			Cascade(m_Wheel.tv5, INDEX(3));
		}

		m_Wheel.curTick++;

		auto entry = m_Wheel.tv1 + idx;
		auto currTimer = entry->head;
		if (nullptr != currTimer)
		{
			m_Wheel.runningTimer = currTimer;
			//清空当前槽
			entry->head = entry->tail = nullptr;
			do
			{
				auto next = currTimer->next;
				currTimer->fun(currTimer->data);
				if (currTimer->runNum > 0)
				{
					currTimer->runNum--;
				}
				if (0 == currTimer->runNum)
				{
					DelTimer(currTimer);//调用完就删除
				}
				else
				{
					//会延时一毫秒运行
					//秒为tick会晚一秒 减一毫秒就不会晚一秒
					auto nowTime = SysTime::GetSysMilSecond() - 1;
					currTimer->expire = nowTime + currTimer->interval;
					AddTimer(currTimer);
				}

				m_Wheel.runningTimer = currTimer = next;
			} while (nullptr != currTimer);
		}
	}
}

uint32_t WTimer::GetNewTimerId()
{
	m_LastTimerId++;
	if (m_LastTimerId <= m_CacheTimerNum)
	{
		m_LastTimerId = m_CacheTimerNum + 1;
	}
	while (m_TimerMap.find(m_LastTimerId) != m_TimerMap.end())
	{
		m_LastTimerId++;
		if (m_LastTimerId <= m_CacheTimerNum)
		{
			m_LastTimerId = m_CacheTimerNum + 1;
		}
	}
	return m_LastTimerId;
}

//真正删除Timer
void WTimer::DelTimer(Timer* timer)
{
	m_TimerMap.erase(timer->id);

	if (timer->id > m_CacheTimerNum)
	{
		SAFE_DELETE(timer);
	}
	else
	{
		m_CustomStack.Push(timer);
	}
}


void WTimer::DelTimer(uint32_t id)
{
	if (nullptr != m_Wheel.runningTimer)
	{
		if (m_Wheel.runningTimer->id == id)
		{
			//调用完后 会减1 为0删除
			m_Wheel.runningTimer->runNum = 1;
			return;
		}
	}

	auto iter = m_TimerMap.find(id);
	if (iter != m_TimerMap.end())
	{
		//DelTimer(iter->second);
		auto timer = iter->second;
		//非链表头部
		if (nullptr != timer->prev)
		{
			timer->prev->next = timer->next;
		}
		else
		{
			auto entry = timer->entry;
			entry->head = timer->next;
			if (timer->next == nullptr)
			{
				entry->tail = nullptr;
			}
		}
		DelTimer(timer);
	}
}

bool WTimer::InitCacheTimer(uint16_t cacheNum)
{
	m_CacheTimerArray = new(nothrow) Timer[cacheNum];
	if (nullptr == m_CacheTimerArray)
	{
		LogError("not enough memory");
		return  false;
	}

	if (!m_CustomStack.Init(cacheNum))
	{
		LogError("not enough memory");
		return  false;
	}

	for (auto i = 0; i < cacheNum; i++)
	{
		m_CacheTimerArray[i].id = i + 1;
		m_CustomStack.Push(&m_CacheTimerArray[i]);
	}
	return true;
}

File: mini_utils/src/wtimer.rs
use crate::time;
use std::collections::VecDeque;
use std::mem::{self, MaybeUninit};

//第1个轮子占用8位
const TVR_BITS: u64 = 8;
//第2~5轮子各占用6位
const TVN_BITS: u64 = 6;

/// 第1个轮子槽数量为256
const TVR_SIZE: u64 = 1 << TVR_BITS;

/// 第2~5轮子槽数量为64
const TVN_SIZE: u64 = 1 << TVN_BITS;

/// 第1个轮子槽0~255
const TVR_MSAK: u64 = TVR_SIZE - 1;

/// 第2~5轮子槽0~63
const TVN_MSAK: u64 = TVN_SIZE - 1;

macro_rules! tv_idx {
    ($tick:expr, $idx:expr) => {
        (($tick >> (TVR_BITS + ($idx) * TVN_BITS)) & TVN_MSAK) as usize
    };
}

macro_rules! cascade {
    ($wheel:expr, $tv:ident, $idx:expr) => {{
        let mut deque = mem::replace(&mut $wheel.$tv[$idx], VecDeque::new());
        while let Some(entity) = deque.pop_front() {
            push_entity($wheel, entity);
        }
        $idx > 0
    }};
}

pub struct WTimer {
    wheel: Wheel,
}
pub trait IWTask {
    /// return true:任务完成
    fn execute(&mut self) -> bool;
}

pub struct TEntity {
    /// 过期时间
    pub(crate) expire: u64,
    /// 运行间隔
    pub(crate) interval: u64,
    /// 任务结构体
    pub(crate) task: Box<dyn IWTask>,
}

struct Wheel {
    cur_tick: u64,
    /// Tick间隔毫秒
    tick_size: u64,
    //cur_entity:TEntity,
    tv1: [VecDeque<TEntity>; TVR_SIZE as usize],
    tv2: [VecDeque<TEntity>; TVN_SIZE as usize],
    tv3: [VecDeque<TEntity>; TVN_SIZE as usize],
    tv4: [VecDeque<TEntity>; TVN_SIZE as usize],
    tv5: [VecDeque<TEntity>; TVN_SIZE as usize],
}

impl WTimer {
    /// tick_size: tick间隔 单位(毫秒)
    pub fn new(tick_size: u16) -> Self {
        let tick_size: u64 = if tick_size < 1 { 1 } else { tick_size as u64 };
        let cur_tick = time::timestamp() / tick_size;

        let mut wheel = Wheel {
            cur_tick,
            tick_size,
            tv1: unsafe { MaybeUninit::uninit().assume_init() },
            tv2: unsafe { MaybeUninit::uninit().assume_init() },
            tv3: unsafe { MaybeUninit::uninit().assume_init() },
            tv4: unsafe { MaybeUninit::uninit().assume_init() },
            tv5: unsafe { MaybeUninit::uninit().assume_init() },
        };

        for i in 0..TVR_SIZE as usize {
            wheel.tv1[i] = VecDeque::new();
        }

        for i in 0..TVN_SIZE as usize {
            wheel.tv2[i] = VecDeque::new();
            wheel.tv3[i] = VecDeque::new();
            wheel.tv4[i] = VecDeque::new();
            wheel.tv5[i] = VecDeque::new();
        }

        WTimer { wheel }
    }
    pub fn scheduled(&mut self, timestamp: u64) {
        let wheel = &mut self.wheel;
        let mut w_tick = wheel.cur_tick;
        let cur_tick = timestamp / wheel.tick_size;
        while w_tick < cur_tick {
            let idx = w_tick & TVR_MSAK;
            if idx != 0
                && !cascade!(wheel, tv2, tv_idx!(w_tick, 0))
                && !cascade!(wheel, tv3, tv_idx!(w_tick, 1))
                && !cascade!(wheel, tv4, tv_idx!(w_tick, 2))
            {
                cascade!(wheel, tv5, tv_idx!(w_tick, 3));
            }

            w_tick += 1;
            wheel.cur_tick = w_tick;
            let mut deque = mem::replace(&mut wheel.tv1[idx as usize], VecDeque::new());
            while let Some(mut entity) = deque.pop_front() {
                if entity.task.execute() {
                    continue;
                }
                entity.expire = timestamp + entity.interval;
                push_entity(wheel, entity);
            }
        }
    }

    pub fn push_task(&mut self, delay: u64, interval: u64, task: Box<dyn IWTask>) {
        let interval = if interval < self.wheel.tick_size {
            self.wheel.tick_size
        } else {
            let n = interval % self.wheel.tick_size;
            if n == 0 {
                interval
            } else {
                interval + (self.wheel.tick_size - n)
            }
        };

        push_entity(
            &mut self.wheel,
            TEntity {
                task,
                interval,
                expire: time::timestamp() + delay,
            },
        );
    }
}

fn push_entity(wheel: &mut Wheel, entity: TEntity) {
    let mut ticks = entity.expire / wheel.tick_size;
    let mut idx = (ticks - wheel.cur_tick) as u64;

    let entity_deque: &mut VecDeque<TEntity>;
    if idx < TVR_SIZE {
        let i = ticks & TVR_MSAK;
        entity_deque = &mut wheel.tv1[i as usize];
    } else if idx < 1 << (TVR_BITS + TVN_BITS) {
        let i = (ticks >> TVR_BITS) & TVN_MSAK;
        entity_deque = &mut wheel.tv2[i as usize];
    } else if idx < 1 << (TVR_BITS + 2 * TVN_BITS) {
        let i = (ticks >> (TVR_BITS + TVN_BITS)) & TVN_MSAK;
        entity_deque = &mut wheel.tv3[i as usize];
    } else if idx < 1 << (TVR_BITS + 3 * TVN_BITS) {
        let i = (ticks >> (TVR_BITS + 2 * TVN_BITS)) & TVN_MSAK;
        entity_deque = &mut wheel.tv4[i as usize];
    }
    /*else if idx < 0 {
        entity_deque = &mut self.wheel.tv1[(self.wheel.cur_tick & TVR_MSAK) as usize];

    }*/
    else {
        if idx > 0xffffffffu64 {
            idx = 0xffffffffu64;
            ticks = idx + wheel.cur_tick;
        }
        let i = (ticks >> (TVR_BITS + 3 * TVN_BITS)) & TVN_MSAK;
        entity_deque = &mut wheel.tv5[i as usize];
    }
    entity_deque.push_back(entity);
}

pub struct TestIWTask {
    pub id: u64,
    pub name: String,
}

impl IWTask for TestIWTask {
    fn execute(&mut self) -> bool {
        self.id += 1;
        if self.id == 1 || self.id % 600 == 0 {
            println!(
                "time:{} id:{} name:{}",
                time::timestamp(),
                self.id,
                self.name
            );
        }
        self.id > 9999999999999
    }
}
#[cfg(test)]
mod test {
    use crate::time;
    use crate::wtimer::TestIWTask;
    use crate::wtimer::WTimer;

    #[test]
    fn test_timer() {
        let mut wtimer = WTimer::new(1);

        for i in 0..9 {
            let task = Box::new(TestIWTask {
                id: 0,
                name: format!("name:{}", i),
            });
            wtimer.push_task(1, 10, task);
        }

        wtimer.scheduled(time::timestamp());

        println!("test_timer finish");
    }
}
